### PR TITLE
Button Core Component

### PR DESCRIPTION
**Pull request by Pareesh**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes and the overall coverage did not decrease.
- [ ] All unit tests pass on CircleCi.
- [ ] I ran all tests locally and they pass.

### CODE REVIEW

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+### Use Object
+The button component uses the `com.adobe.cq.forms.core.components.models.form.Button` Sling Model for its Use-object.
+
+### Edit Dialog Properties
+The following properties are written to JCR for this button component and are expected to be available as `Resource` properties:
+
+1. `./jcr:title` - defines the label to use for this field
+2. `./hideTitle` - if set to `true`, the label of this field will be hidden
+3. `./name` - defines the name of the field, which will be submitted with the form data
+4. `./default` - defines the default value of the field
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md:38]: are we capturing default ?

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+JavaScript handling for dialog interaction. It is already included by its edit dialog.
+
+## BEM Description
+```
+BLOCK cmp-button
+    ELEMENT cmp-button__text
+    ELEMENT cmp-button__questionmark
+    ELEMENT cmp-button__shortdescription
+    ELEMENT cmp-button__longdescription
+```
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md:55]: Does button has a long description ?

Comment (vdua) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md:55]: I think we should add icon support in button

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/_cq_dialog/.content.xml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                        jcr:title="Basic"
+                        sling:resourceType="granite/ui/components/coral/foundation/container">
+                        <items 
+                          jcr:primaryType="nt:unstructured"
+                          sling:hideChildren="[placeholder, hideTitle]">
+                          <default
+                            jcr:primaryType="nt:unstructured"
+                            sling:resourceType="granite/ui/components/coral/foundation/form/textfield"
+                            fieldLabel="Button Default Value"
+                            name="./default"/>
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/_cq_dialog/.content.xml:46]: Either default or value but do not capture both in the dialog

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/button.html [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  data-sly-use.longDescription="${renderer.longDescriptionPath}"
+  data-sly-use.questionMark="${renderer.questionMarkPath}"></sly>
+<button 
+  type="${button.buttonType}"
+  value="${button.value}"
+  id="${component.id}"
+  class="cmp-button"
+  data-cmp-is="adaptiveFormButton"
+  aria-label="${button.label.value}">
+<span data-sly-test="${button.label.value}" class="cmp-button__text">${button.label.value}</span>
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/button.html:30]: cp-button__label. Can you check the wcm button and align the markup with that
cc @nit23uec @rismehta

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                    this.element.closest("form").submit();
+                }
+            } else {
+                onClick = (event) => {
+                    // code for click
+                }
+            }
+            this.element.addEventListener("click", function() {
+                console.log("hello")
+            });
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js:78]: This is not supposed to work. The click handlers are defined in the form model.

**Review by vdua — CHANGES_REQUESTED** [review]

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/editor/js/editDialog.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+/*******************************************************************************
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/editor/js/editDialog.js:1]: pls remove if not required.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md (hunk 1 above)
Comment (vdua) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md:38]: Let's not capture default.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md (hunk 2 above)
Comment (vdua) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md:55]: We decided to add the support later.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+5. `./description` - defines a help message that can be rendered in the field as a hint for the user
+
+## Client Libraries
+The component provides a `core.forms.components.button.v1` client library category that contains a JavaScript
+component. It should be added to a relevant site client library using the `embed` property.
+
+It also provides a `core.forms.components.button.v1.editor` editor client library category that includes
+JavaScript handling for dialog interaction. It is already included by its edit dialog.
+
+## BEM Description
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md:48]: Please update the BEM as per the html markup

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/_cq_dialog/.content.xml [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+                  fieldDescription="Value of button"
+                  granite:hide="{Boolean}true"
+                  name="./value"/>
+                <default
+                  jcr:primaryType="nt:unstructured"
+                  sling:resourceType="granite/ui/components/coral/foundation/form/textfield"
+                  fieldLabel="Default"
+                  fieldDescription="Default value of button"
+                  granite:hide="{Boolean}true"
+                  name="./default"/>
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/_cq_dialog/.content.xml:55]: let's not capture default

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js (hunk 5 above)
Comment (vdua) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js:58]: `this._model.dispatchEvent(new FormView.Click())`

@nit23uec Have we exposed these objects ?

**Review by vdua — CHANGES_REQUESTED** [review]

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/_cq_dialog/.content.xml (hunk 8 above)
Comment (Pareesh) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/_cq_dialog/.content.xml:54]: @vdua  it is hidden, it won't display on ui. I will remove it if not required at all.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    ELEMENT cmp-button__text
+    ELEMENT cmp-button__questionmark
+    ELEMENT cmp-button__shortdescription
+    ELEMENT cmp-button__longdescription
+```
+
+## JavaScript Data Attribute Bindings
+
+Apply a `data-cmp-is="adaptiveFormButton"` attribute to the button to enable initialization of the JavaScript component.
+
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/README.md:51]: pls add adaptive form container path also in the list of attributes.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            text: `.${Button.bemBlock}__text`,
+        };
+
+        /**
+         * Return the questionmark element.
+         * @readonly
+         * @returns {HTMLElement}
+         */
+        get qm() {
+            return document.getElementById(`${this.element.id}-questionMark`);
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js:41]: pls have the constants in static selectors block.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+            self: "[data-" + this.NS + '-is="' + this.IS + '"]',
+            text: `.${Button.bemBlock}__text`,
+        };
+
+        /**
+         * Return the questionmark element.
+         * @readonly
+         * @returns {HTMLElement}
+         */
+        get qm() {
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js:40]: get qm?

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js (hunk 5 above)
Comment (nit23uec) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/clientlibs/site/js/buttonview.js:73]: no, pls use this._model.dispatchEvent({type: "click"})

**Pull request merged**